Improve query sort options and implementations

apmoriarty opened the pull request:
- consolidate and rename variables that control sorting query pre and post global index lookup
- expanded test cases for both pre and post index sorting
- reworked sort implementations

apmoriarty committed:
consolidate and rename variables that control sorting query pre and post global index lookup

apmoriarty committed:
Merge branch 'integration' into task/consolidate-sort-options

apmoriarty committed:
Merge branch 'integration' into task/consolidate-sort-options

jwomeara reviewed (changes requested):
From a serialization perspective, your changes to ShardQueryConfiguration are not backwards compatible.  We serialize the ShardQueryConfiguration to hazelcast for the query microservices, so I would avoid making changes like this (ones that have no impact on functionality) in the future as it is likely to break query microservices.  However, since you made a change, please update ShardQueryConfiguration's serialVersionUID.  If you leave that unchanged we will run into serialization errors in the query microservices.

apmoriarty committed:
Update ShardQueryConfiguration serialVersionUID

jwomeara reviewed (dismissed):
with the addition of the serialVersionUID update, this should be ok for query microservices.  not sure what renaming all of these buys us though...

hgklohr committed:
Merge branch 'integration' into task/consolidate-sort-options

apmoriarty committed:
Replaced complex FieldOrTermNodeCostComparator with specific Field or Term comparators

Fixed bug in IndexInfo that double counted the first node

Fixed bug in RangeStream that used the wrong count

Fixed bug in NodeCostComparator that led to an integer overflow

Added integration tests for ordering visitiors in RangeStream and ShapesTest

ivakegg commented:
I had previously approved this, but since then this PR is showing a heck of a lot more changes than just the variable renaming.  What happened here?

apmoriarty commented:
> I had previously approved this, but since then this PR is showing a heck of a lot more changes than just the variable renaming. What happened here?

Someone requested additional tests which exposed several cases where sorting was not behavior according to expectation.

apmoriarty committed:
Merge branch 'integration' into task/consolidate-sort-options

ivakegg commented:
> > I had previously approved this, but since then this PR is showing a heck of a lot more changes than just the variable renaming. What happened here?
> 
> Someone requested additional tests which exposed several cases where sorting was not behavior according to expectation.

Well you also added additional sorting mechanisms as well, not just test cases.

apmoriarty commented:
> Well you also added additional sorting mechanisms as well, not just test cases.

I have updated the title and description to reflect that this was not just an update to some test cases

ivakegg commented on warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java:
```diff
@@ -2763,8 +2763,10 @@ public Tuple2<CloseableIterable<QueryPlan>,Boolean> getQueryRanges(ScannerFactor
+        if (config.isSortQueryPreIndexWithFieldCounts()) {
             config.setQueryTree(timedSortQueryBeforeGlobalIndex(config, getMetadataHelper()));
+        } else if (config.isSortQueryPreIndexWithImpliedCounts()) {
+            config.setQueryTree(OrderByCostVisitor.order(config.getQueryTree()));
```
Should this second ordering option be timed as well?

apmoriarty replied:
Yes, yes it should. Thanks!

apmoriarty committed:
Add timed method for sorting by implied counts

apmoriarty committed:
Merge branch 'integration' into task/consolidate-sort-options

apmoriarty committed:
Merge branch 'integration' into task/consolidate-sort-options

apmoriarty committed:
Merge branch 'integration' into task/consolidate-sort-options

apmoriarty committed:
Merge branch 'integration' into task/consolidate-sort-options
